### PR TITLE
Add SQL API

### PR DESCRIPTION
**Pull request by brfrn169**
This PR adds the new SQL API.

The example code is available in the following (It has been deleted in the `add-sql-api` branch):
https://github.com/scalar-labs/scalardb/blob/f99fdbb519c18ad0357bebbb3e860980d0ef62a2/core/src/main/java/com/scalar/db/sql/Example.java

Please take a look!

### CODE REVIEW

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+  private static final String COLUMN_NAME_4 = "col4";
+
+  public static void main(String[] args) {
+    try (SqlSessionFactory sqlSessionFactory =
+        SqlSessionFactory.builder()
+            .withProperty("scalar.db.contact_points", "jdbc:mysql://localhost:3306/")
+            .withProperty("scalar.db.username", "root")
+            .withProperty("scalar.db.password", "mysql")
+            .withProperty("scalar.db.storage", "jdbc")
+            .build()) {
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:33]: We fist need to create this `SqlSessionFactory`. I'm going to add some more utility methods to the builder class of `SqlSessionFactory`.

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+      storageModeExample(sqlSessionFactory);
+      transactionModeExample(sqlSessionFactory);
+      twoPhaseCommitTransactionModeExample(sqlSessionFactory);
+    }
+  }
+
+  public static void storageModeExample(SqlSessionFactory sqlSessionFactory) {
+    System.out.println("Start storage mode example");
+
+    StorageSqlSession storageSqlSession = sqlSessionFactory.getStorageSqlSession();
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:43]: We use `StorageSqlSession` for `Storage` mode.

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+
+    ResultSet resultSet;
+
+    // Create namespace
+    CreateNamespaceStatement createNamespaceStatement =
+        SqlBuilder.createNamespace(NAMESPACE_NAME).ifNotExists().build();
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:49]: All statements are created with this `SqlBuilder` and I designed this API looks SQL as much as possible.

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+            .ifNotExists()
+            .withPartitionKey(Attribute.ID, DataType.TEXT)
+            .withColumn(Attribute.STATE, DataType.INT)
+            .withColumn(Attribute.CREATED_AT, DataType.BIGINT)
+            .build();
+    resultSet = storageSqlSession.execute(createTableStatement);
+    resultSet.close();
+
+    // Begin
+    TransactionSqlSession transactionSqlSession = sqlSessionFactory.beginTransaction();
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:223]: We use `TransactionSqlSession ` for Transaction mode.

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+    resultSet = storageSqlSession.execute(createTableStatement);
+    resultSet.close();
+
+    // Begin
+    TwoPhaseCommitTransactionSqlSession twoPhaseCommitTransactionSqlSession1 =
+        sqlSessionFactory.beginTwoPhaseCommitTransaction();
+
+    TwoPhaseCommitTransactionSqlSession twoPhaseCommitTransactionSqlSession2 =
+        sqlSessionFactory.joinTwoPhaseCommitTransaction(
+            twoPhaseCommitTransactionSqlSession1.getTransactionId());
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:360]: We use `TwoPhaseCommitTransactionSqlSession` for Two-phase Commit Transaction mode.

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+    CreateTableStatement createTableStatement =
+        SqlBuilder.createTable(NAMESPACE_NAME, TABLE_NAME)
+            .ifNotExists()
+            .withPartitionKey(COLUMN_NAME_1, DataType.INT)
+            .withClusteringKey(COLUMN_NAME_2, DataType.TEXT)
+            .withColumn(COLUMN_NAME_3, DataType.BIGINT)
+            .withColumn(COLUMN_NAME_4, DataType.FLOAT)
+            .withClusteringOrder(COLUMN_NAME_2, Order.ASC)
+            .build();
+    resultSet = storageSqlSession.execute(createTableStatement);
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:63]: We execute the statement with this `execute` method and it returns `ResultSet`.

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+    resultSet.close();
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:64]: We need to close `ResultSet` after using it.

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+        SqlBuilder.insertInto(NAMESPACE_NAME, TABLE_NAME)
+            .values(
+                Assignment.column(COLUMN_NAME_1).value(Value.ofInt(10)),
+                Assignment.column(COLUMN_NAME_2).value(Value.ofText("abc")),
+                Assignment.column(COLUMN_NAME_3).value(Value.ofBigInt(100L)),
+                Assignment.column(COLUMN_NAME_4).value(Value.ofFloat(1.23F)))
+            .ifNotExists()
+            .build();
+    resultSet = storageSqlSession.execute(insertStatement);
+    System.out.println("InsertStatement result: " + resultSet.one().get().getBoolean(0));
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:77]: Conditional mutations return the result from `ResultSet`.

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+            .build();
+
+    try (ResultSet rs = storageSqlSession.execute(selectStatement)) {
+      for (Record record : rs) {
+        System.out.println("column1 value: " + record.getInt(COLUMN_NAME_1));
+        System.out.println("column2 value: " + record.getText(COLUMN_NAME_2));
+        System.out.println("column3 value: " + record.getBigInt(COLUMN_NAME_3));
+        System.out.println();
+      }
+    }
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:136]: We can also get the Select statement results from `ResultSet`.

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+  public static void main(String[] args) {
+    try (SessionFactory sessionFactory =
+        SessionFactory.builder()
+            .withProperty("scalar.db.contact_points", "jdbc:mysql://localhost:3306/")
+            .withProperty("scalar.db.username", "root")
+            .withProperty("scalar.db.password", "mysql")
+            .withProperty("scalar.db.storage", "jdbc")
+            .build()) {
+      transactionModeExample(sessionFactory);
+      twoPhaseCommitTransactionModeExample(sessionFactory);
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:24]: I removed the storage mode.

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+    }
+  }
+
+  public static void transactionModeExample(SessionFactory sessionFactory) {
+    System.out.println("Start transaction mode example");
+
+    Session session = sessionFactory.getTransactionSession();
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:31]: We can get a session for transaction mode from `SessionFactory.getTransactionSession()`. And we can repeatedly reuse this session after committing or rolling back a transaction.

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+          StatementBuilder.update(NAMESPACE_NAME, TABLE_NAME)
+              .set(
+                  Assignment.column(COLUMN_NAME_3).value(Value.ofBigInt(200L)),
+                  Assignment.column(COLUMN_NAME_4).value(Value.ofFloat(4.56F)))
+              .where(Condition.column(COLUMN_NAME_1).isEqualTo(Value.ofInt(10)))
+              .and(Condition.column(COLUMN_NAME_2).isEqualTo(Value.ofText("abc")))
+              .build());
+
+      // Select
+      ResultSet resultSet =
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:83]: We no longer need to close `ResultSet`

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+          session.executeQuery(
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:84]: We can use `executeQuery()` for Select queries and it returns `ResultSet`.

File: core/src/main/java/com/scalar/db/sql/Example.java [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+
+    // Drop a coordinator table
+    session.execute(StatementBuilder.dropCoordinatorTable().ifExists().build());
+  }
+
+  public static void twoPhaseCommitTransactionModeExample(SessionFactory sessionFactory) {
+    System.out.println("Start two-phase commit transaction mode example");
+
+    Session session1 = sessionFactory.getTwoPhaseCommitTransactionSession();
+    Session session2 = sessionFactory.getTwoPhaseCommitTransactionSession();
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/sql/Example.java:138]: We can get a session for transaction mode from `SessionFactory.getTwoPhaseCommitTransactionSession()`. And we can also repeatedly reuse this session after committing or rolling back a transaction.

**Comment by brfrn169** [comment]
@feeblefakie I changed the API based on the review. Please take a look again!
https://github.com/scalar-labs/scalardb/blob/add-sql-api/core/src/main/java/com/scalar/db/sql/Example.java

**Comment by brfrn169** [comment]
@feeblefakie This PR is ready to be merged. Please take a look!

**Review by feeblefakie — APPROVED** [review]
LGTM! Thank you!
I left one comment but it doesn't block from it being merged.

File: core/src/main/java/com/scalar/db/sql/TransactionSqlSession.java [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+  public ResultSet executeQuery(SelectStatement statement) {
+    checkIfTransactionBegun();
+
+    statementValidator.validate(statement);
+    resultSet = dmlStatementExecutor.execute(transaction, statement);
+    return resultSet;
+  }
+
+  @Override
+  public void prepare() {
```
[the remark this answers is not in the document]
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/sql/TransactionSqlSession.java:116]: > should SqlSession have only the common methods between TransactionSqlSession and TwoPhaseCommitTransactionSqlSession?

Thank you for pointing it out!
Honestly, I'm also not sure which is better. I'm working on another project using this new API now, and if I find that having only the common methods in `SqlSession` is better, I will create a new PR to change it. 
I'll merge it for now. Thank you!

**Pull request merged**